src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -70,589 +91,522 @@ public async ValueTask DisposeAsync()
     }
 
-    public ValueTask SubscribeAsync(int sid, string subject, string? queueGroup, int? maxMsgs, CancellationToken cancellationToken)
+    public async ValueTask UnsubscribeAsync(int sid, int? maxMsgs, CancellationToken cancellationToken)
     {
-#pragma warning disable CA2016
-#pragma warning disable VSTHRD103
-        if (!_semLock.Wait(0))
-#pragma warning restore VSTHRD103
-#pragma warning restore CA2016
-        {
-            return SubscribeStateMachineAsync(false, sid, subject, queueGroup, maxMsgs, cancellationToken);
-        }
-
-        if (_flushTask is { IsCompletedSuccessfully: false })
-        {
-            return SubscribeStateMachineAsync(true, sid, subject, queueGroup, maxMsgs, cancellationToken);
-        }
-
+        Interlocked.Add(ref _counter.PendingMessages, 1);
+        await _semLock.WaitAsync(cancellationToken).ConfigureAwait(false);
         try
         {
             if (_disposed)
             {
                 throw new ObjectDisposedException(nameof(CommandWriter));
             }
 
-            var success = false;
-            try
-            {
-                _protocolWriter.WriteSubscribe(sid, subject, queueGroup, maxMsgs);
-                success = true;
-            }
-            finally
+            PipeWriter bw;
+            lock (_lock)
             {
-                EnqueueCommand(success);
+                if (_pipeWriter == null)
+                    ThrowOnDisconnected();
+                bw = _pipeWriter!;
             }
+
+            _protocolWriter.WriteUnsubscribe(bw, sid, maxMsgs);
+            await bw.FlushAsync(cancellationToken).ConfigureAwait(false);
         }
         finally
         {
             _semLock.Release();
+            Interlocked.Add(ref _counter.PendingMessages, -1);
         }
-
-        return ValueTask.CompletedTask;
     }
 
-    public ValueTask UnsubscribeAsync(int sid, CancellationToken cancellationToken)
-    {
-#pragma warning disable CA2016
-#pragma warning disable VSTHRD103
-        if (!_semLock.Wait(0))
-#pragma warning restore VSTHRD103
-#pragma warning restore CA2016
-        {
-            return UnsubscribeStateMachineAsync(false, sid, cancellationToken);
-        }
-
-        if (_flushTask is { IsCompletedSuccessfully: false })
-        {
-            return UnsubscribeStateMachineAsync(true, sid, cancellationToken);
-        }
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static void ThrowOnDisconnected() => throw new NatsException("Connection hasn't been established yet.");
 
+    [AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder))]
+    private async ValueTask PublishLockedAsync(string subject, string? replyTo,  NatsPooledBufferWriter<byte> payloadBuffer, NatsPooledBufferWriter<byte>? headersBuffer, CancellationToken cancellationToken)
+    {
+        // Interlocked.Add(ref _counter.PendingMessages, 1);
+        await _semLock.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Speaking of, putting the `WaitAsync` is probably 'better' here overall (the old `Wait(0)` -> statemachine pattern has some thrashing concerns) but it does raise some interesting questions in my head again as to the driver.


(Below is a language-lawyer disclaimer and... well, hopefully we aren't worried about it, but if we are, wanted to bring it up.)
This is going to be more fair than the old method (b/c it's always `WaitAsync`, but it is worth noting that `SemaphoreSlim` like most locking primitives in .NET do not guarantee fairness as to which thread gets released first [citation](https://github.com/dotnet/runtime/issues/19336#issuecomment-259982210). 

If we -do- need stronger ordering/fairness on these, we will need to consider an alternative... and also remember the caveats mentioned in the citation; they are certainly not 'absolutes' however they may require consideration in a solution (again, only if needed, this can be a rabbit hole.)

Reply:
just swapped this out with a channel. WDYT? (@stebet also raised concerns about `SemaphoreSlim` before)